### PR TITLE
Make `JoinKeys` related changes due to refactor

### PR DESCRIPTION
**Pull request by vedhav**
Related to [teal.data PR #184](https://github.com/insightsengineering/teal.data/pull/184)
Make changes to `teal.slice` because of the refactor to the `JoinKeys` class from R6 to S3.

This diagram shows the R6 methods along with the replacement S3 methods/functions.

Screenshot 2023-11-15 at 3 56 36 PM

### CODE REVIEW

**Comment by m7pr** [comment]
Wow, great diagram, very heplful!

**Comment by pawelru** [comment]
Please vbump teal.data deps once you got it merged and new version number available

**Comment by m7pr** [comment]
@vedhav is the diagram still up to date? I think at least we changed JoinKeys name to join_keys : ) This was great help and I reckon we could keep it updated

**Comment by vedhav** [comment]
@m7pr updated the diagram

**Comment by m7pr** [comment]
thanks, you da man

File: NEWS.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 * Specified minimal version of package dependencies.
 * Removed storing and restoring of `teal_slices` objects.
+* Update documentation and code to reflect `teal.data::JoinKeys` class and logic change to `teal.data::join_keys`.
```
Comment (kartikeyakirar) [inline NEWS.md:7]: do you mean refactor or remove?
```suggestion
* Update documentation and code to refactor/remove `teal.data::JoinKeys` class and logic change to `teal.data::join_keys`.
```

File: R/FilteredData.R [hunk 2]
```diff
@@ -333,24 +332,28 @@ FilteredData <- R6::R6Class( # nolint
       # the UI also uses `datanames` in ids, so no whitespaces allowed
       check_simple_name(dataname)
 
-      join_keys <- self$get_join_keys()
-      parent_dataname <- join_keys$get_parent(dataname)
+      parent_dataname <- teal.data::parent(private$join_keys, dataname)
+      keys <- private$join_keys[[dataname]][[dataname]]
```
Comment (kartikeyakirar) [inline R/FilteredData.R:336]:
```suggestion
      keys <- private$join_keys[dataname, dataname]
```

File: R/FilteredData.R [hunk 3]
```diff
@@ -333,24 +332,28 @@ FilteredData <- R6::R6Class( # nolint
...
         private$filtered_datasets[[dataname]] <- init_filtered_dataset(
           dataset = data,
           dataname = dataname,
           metadata = metadata,
           label = label,
-          keys = self$get_join_keys()$get(dataname, dataname)
+          keys = keys
         )
       } else {
+        join_keys <- private$join_keys[[dataname]][[parent_dataname]]
```
Comment (kartikeyakirar) [inline R/FilteredData.R:348]:
```suggestion
        join_keys <- private$join_keys[dataname, parent_dataname]
```

File: NEWS.md (hunk 1 above)
Comment (vedhav) [reply NEWS.md:7]: How about this?

```suggestion
* Update documentation and code to reflect the changes due to the refactor of `teal.data::JoinKeys` into `teal.data::join_keys`.
```

Comment (kartikeyakirar) [reply NEWS.md]: its  better

**Review by kartikeyakirar — APPROVED** [review]
changes looks good!

**Pull request merged**